lib/zip_source_zip_new.c

```diff
@@ -43,20 +43,43 @@ zip_source_t *_zip_source_zip_new(zip_t *srcza, zip_uint64_t srcidx, zip_flags_t
     zip_stat_t st;
     zip_file_attributes_t attributes;
     zip_dirent_t *de;
-    bool partial_data, needs_crc, needs_decrypt, needs_decompress;
+    bool partial_data, needs_crc, encrypted, needs_decrypt, compressed, needs_decompress, reopening;
+    zip_flags_t stat_flags;
+    zip_t *source_archive;
+    zip_uint64_t source_index;
 
     if (srcza == NULL || srcidx >= srcza->nentry || len > ZIP_INT64_MAX) {
         zip_error_set(error, ZIP_ER_INVAL, 0);
         return NULL;
     }
 
-    if ((flags & ZIP_FL_UNCHANGED) == 0 && (ZIP_ENTRY_DATA_CHANGED(srcza->entry + srcidx) || srcza->entry[srcidx].deleted)) {
-        zip_error_set(error, ZIP_ER_CHANGED, 0);
-        return NULL;
+    if ((flags & ZIP_FL_UNCHANGED) == 0) {
+        zip_entry_t *entry = srcza->entry + srcidx;
+        if (ZIP_ENTRY_DATA_CHANGED(entry)) {
+            if (!zip_source_supports_reopen(entry->source)) {
+                zip_error_set(error, ZIP_ER_CHANGED, 0);
+                return NULL;
+            }
+
+            reopening = true;
+        } else if (entry->deleted) {
+            zip_error_set(error, ZIP_ER_CHANGED, 0);
+            return NULL;
+        }
+    }
+    else {
+        reopening = false;
+    }
+
+    stat_flags = flags;
+    if (!reopening) {
+        stat_flags |= ZIP_FL_UNCHANGED;
     }
 
-    if (zip_stat_index(srcza, srcidx, flags | ZIP_FL_UNCHANGED, &st) < 0) {
-        zip_error_set(error, ZIP_ER_INTERNAL, 0);
+    if (zip_stat_index(srcza, srcidx, stat_flags, &st) < 0) {
+        /* if reopening, the error could come from 3rd party code, so it's not internal. */
+        int ze = (reopening ? ZIP_ER_INVAL : ZIP_ER_INTERNAL);
+        zip_error_set(error, ze, 0);
```

Comment:
It's probably best to just return the error from the source unchanged. (Also, ZIP_ER_INVAL is for invalid arguments, so doesn't fit here.)